CRAB-34138: Create Excel Spreadsheet with Resources and Upload to S3

alizbay94 opened the pull request:
This PR allows NagBot to create a formatted excel spreadsheet containing the data pertaining to each of the Resources it processes, and the spreadsheet includes a worksheet for each resource. The spreadsheet is generated during the `notify` command, and it is created using the `xlsxwriter` library. 

The spreadsheet is uploaded to the [nagbot-spreadsheets](https://s3.console.aws.amazon.com/s3/buckets/nagbot-spreadhseets?region=us-east-1&tab=objects) S3 bucket in the main Seeq AWS account, and the format for naming the spreadsheets is `YYYY-MM-DD-NagBot-Report.xlsx`.

Here is an example spreadsheet that was produced: 
[2022-11-08-NagBot-Report.xlsx](https://github.com/seeq12/nagbot/files/9963705/2022-11-08-NagBot-Report.xlsx)

I've verified the changes work when running in a Docker container, and on my machine.

One thing I was unsure of:
Do we need to cleanup the excel file after uploading? I think the file should be deleted at the conclusion of docker run when the container dies, but I put logic there to remove the file in case.

alizbay94 committed:
Change ownershipe of /usr/src so user can write spreadsheet to that directory

alizbay94 committed:
Create excel spreadsheet and add worksheet for each resource

alizbay94 committed:
Merge branch 'master' into task/azb/write-resources-to-spreadsheet-and-upload-to-s3-CRAB-34138

alizbay94 committed:
Increment NagBot version

alizbay94 commented on app/spreadsheet.py:
```diff
@@ -0,0 +1,48 @@
+        print(e)
+
+    # Cleanup and delete the workbook
+    os.remove(filename)
```
wasn't sure if I needed this since it will be running in a docker container and file should be gone when container dies, but left there for good measure.

CodyRay commented on app/nagbot.py:
```diff
@@ -28,7 +29,7 @@
 4. Environment variables
    * "SLACK_BOT_TOKEN" containing a token allowing messages to be posted to Slack.
-   * "GDOCS_SERVICE_ACCOUNT_FILENAME" containing the name of the google sheet
+   * "GDOCS_SERVICE_ACCOUNT_FILENAME" containing the name of the google sheet 
```
Is all the Gdocs stuff ripped out? Maybe we can remove these references

alizbay94 replied:
I think its all gone, i'll double check and remove the old references.

CodyRay commented on app/spreadsheet.py:
```diff
@@ -0,0 +1,48 @@
+    worksheet.set_column(0, len(resource_header)+1, 20)
+
+    # Write resources data to worksheet
+    col_num = 1
```
Consider using `enumerate()` https://www.geeksforgeeks.org/enumerate-in-python/

CodyRay commented on app/spreadsheet.py:
```diff
@@ -0,0 +1,48 @@
+    s3_client = boto3.client('s3')
+    bucket = "nagbot-spreadhseets"
+    try:
+        # response = s3_client.upload_file(file_name, bucket, object_name)
```
Commented code could be deleted :recycle:

CodyRay commented on app/nagbot.py:
```diff
@@ -38,9 +39,15 @@ def notify_internal(channel, dryrun):
         summary_msg = f"Hi, I'm Nagbot v{__version__} :wink: "
         summary_msg += "My job is to make sure we don't forget about unwanted AWS resources and waste money!\n"
 
+        filename = f"{TODAY_YYYY_MM_DD}-NagBot-Report.xlsx"
```
Consider using [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory), then python will take care of cleaning up and it is in a location that is generally writable by all users

CodyRay commented on app/nagbot.py:
```diff
@@ -81,6 +88,7 @@ def notify_internal(channel, dryrun):
                 else:
                     summary_msg += f'No {ec2_type}s are due to be stopped at this time.\n'
+        spreadsheet.upload_spreadsheet_to_s3(filename, workbook)
         sqslack.send_message(channel, summary_msg)
```
Can we link the url in the message?

alizbay94 replied:
Good idea, I could set that up.

CodyRay commented on app/spreadsheet.py:
```diff
@@ -0,0 +1,48 @@
+        resource_data = [", ".join(data) if type(data) is list else data for data in resource_data]
+        worksheet.write_row(col_num, 0, resource_data, standard_format)
+        col_num += 1
+    return workbook
```
Doing something like https://stackoverflow.com/questions/29463274/simulate-autofit-column-in-xslxwriter would be nice to have; make it a little bit nicer to use

If we could turn on the [make it a table](https://www.tutorialspoint.com/python_xlsxwriter/python_xlsxwriter_tables.htm) that would be handy too

CodyRay reviewed (commented):
Looking great, some comments an improvements

CodyRay commented on app/spreadsheet.py:
```diff
@@ -0,0 +1,48 @@
+
+    # create header in worksheet
+    bold_format = workbook.add_format({'bold': True, "align": "left", "font_size": 15})
+    standard_format = workbook.add_format({"font_size": 12, 'align': 'left'})
```
I'd vote to remove the font size configuration - seems very big

CodyRay reviewed (commented):
Some further notes after playing with the example workbook:

I noticed some weirdness for Snapshots: they have some columns - OS, IOPS, and Throughput that don't make sense for volumes

Some other ideas for useability improvements:
 * Could the ID columns be clickable links to open the resource in the console?
 * Could we add an "Overview" worksheet that has a summary of spending for each resource type?
 * Could we sort the rows by descending costs? Have the highest cost items front and center

If any of those nice to haves end up being difficult to add, that is ok to skip them

alizbay94 committed:
Remove references to google docs

alizbay94 committed:
Save excel spreadsheet to a temporary directory

alizbay94 committed:
Include AWS keys as necessary env vars in comment in nagbot.py

alizbay94 committed:
Include url to s3 bucket in slack message at conclusion of notify

alizbay94 committed:
Remove font size configuration

alizbay94 committed:
Use enumerate() instead of a counter to iterate numbers in loop

CodyRay commented on app/spreadsheet.py:
```diff
@@ -32,17 +33,19 @@ def add_resource_worksheet_to_workbook(workbook, resources, resource_name):
-    # save the workbook to the cwd by closing
+    cwd = os.getcwd()
+    # save the workbook to a temporary directory
+    temp_directory = tempfile.TemporaryDirectory(ignore_cleanup_errors=True)
```
Does this imply the file isn't actually created until the `close` command is called? If so, I think you can revert the changes to the Dockerfile

CodyRay replied:
Consider [with](https://realpython.com/python-with-statement/) or a try-finally to ensure cleanup

alizbay94 replied:
Implementing a try-finally now

alizbay94 committed:
Include Created by CreateImage as a string to look for in Snapshot description to determine if it is an AMI snapshot

alizbay94 committed:
Optimize speed of checking if AMI is registered by generating a list prior to checks

alizbay94 committed:
Merge changes from master branch

alizbay94 committed:
Create table with resource data for each worksheet, and format table columns

alizbay94 commented:
I've implemented logic to include hyperlinks for each resource id in the excel file that links to the resource on AWS. I've also made the data from each resource worksheet into a table, and wrote custom logic to format the column width of each column based on the longest string in the column including the header. 
I'm thinking I won't sort the table entries in descending order of price since it would require writing custom logic, and we could just use the table to sort the entries in any order we choose when we open the excel file in the Excel UI. There's a drop down next to each column header where we can sort the entries however we choose.

The only feedback i've got left to address in this PR is adding an 'Overview' sheet - @haroldhues, would the idea be that this sheet would essentially contain 4 entries, one entry for the total cost for each of the resources?

alizbay94 committed:
Implement a try finally to ensure cleanup

CodyRay reviewed (approved):
Nice! Great improvements

CodyRay commented on app/snapshot.py:
```diff
@@ -31,10 +32,7 @@ def to_header() -> [str]:
                 'Monthly Price',
```
I found that the spreadsheet doesn't sort the price columns for instances correctly because they contain a literal dollar sign.

![image](https://user-images.githubusercontent.com/5783817/204622450-e580d900-0c3d-46b2-b593-a80c68034d11.png)

I think the contents of the cell needs to be without the dollar sign and then format as "Currency". Then it will appear with the dollar sign, but Excel will do calculations without the dollar sign.

This only seems to apply to instances, the other types don't have the dollar sign.

alizbay94 replied:
I've noticed the same thing, it must be due to the way instance prices are calculated. I'll get them all uniform!

CodyRay commented on app/ami.py:
```diff
@@ -32,9 +32,8 @@ def to_header() -> [str]:
-                'VolumeType'
-                'Throughput',
-                'Snapshot IDs']
+                'VolumeType',
```
I think for AMI we are missing the "AMI name" . Every AMI seems to have it, but I think we are pulling the Name tag. For example, https://us-west-1.console.aws.amazon.com/ec2/v2/home?region=us-west-1#Images:visibility=owned-by-me;search=ami-046531e848f15f786 doesn't show a name in the spreadsheet

Also, do we know if we have creation date available for all resources?

CodyRay commented:
> The only feedback i've got left to address in this PR is adding an 'Overview' sheet - @haroldhues, would the idea be that this sheet would essentially contain 4 entries, one entry for the total cost for each of the resources?

Yep, I think we could potential show other things there in the future as well like a regional breakdown or counts of resources, but total cost for each resource is a good place to start imo

alizbay94 committed:
Change name of s3 bucket where excel file is pushed

alizbay94 committed:
Create summary worksheet for workbook with total cost of all resources

alizbay94 committed:
Include creation timestamp attribute for each resource, and implement logic to retrieve AMI name

alizbay94 commented:
@haroldhue , each of the resources has a creation timestamp, but it is called something different for each resource. I implemented the logic to retrieve the creation timestamp for each resource depending on what it is called, and added an attribute to each resource object with that info, and included a column with that info for each worksheet.
I also implemented the logic to retrieve AMI names, and the AMI names are showing up as expected now 👍🏼
Finally I created an additional 'Summary' worksheet in the workbook with the total cost of each resource.

CodyRay commented:
Latest version is working pretty well - well enough for me to get myself into trouble :cry: 

Is it possible to make the summary the first worksheet?

alizbay94 commented:
Yup! I can get that done.

alizbay94 committed:
Set the summary worksheet as the first visible sheet in the workbook

alizbay94 commented:
I was able to find a method of the `worksheet` class called `activate()` which enables us to set the summary sheet as the first visible sheet when the workbook is open, i've included it in the `add_summary_worksheet_to_workbook` method.

CodyRay commented on app/spreadsheet.py:
```diff
@@ -0,0 +1,85 @@
+        os.chdir(cwd)
+        temp_directory.cleanup()
+
+    return f"https://s3.console.aws.amazon.com/s3/buckets/{bucket}"
```
I think if you add `?prefix={filename}` here then you'll get a link directly to the uploaded file

alizbay94 replied:
I'll add that and test before I merge 👍🏼

CodyRay reviewed (approved):
One last suggestion, but working great - ship it!

alizbay94 committed:
Provide link to spreadsheet and refactor slack message
